src/Surface.cpp

```diff
@@ -798,31 +806,75 @@ int Circle::intersection(Point* point, double angle, Point* points) {
     else if (discr == 0) {
       xcurr = x0;
       ycurr = -b / (2*a);
-      points[num].setCoords(xcurr, ycurr);
-      if (angle < M_PI && ycurr > y0)
-        num++;
-      else if (angle > M_PI && ycurr < y0)
-        num++;
-      return num;
+      zcurr = z0 + sqrt(pow(ycurr - y0, 2.0) + pow(xcurr - x0, 2.0))
+        * tan(M_PI_2 - polar);
+      points[num].setCoords(xcurr, ycurr, zcurr);
+
+      /* Check that point is in same direction as angle */
+      if (azim < M_PI && ycurr > y0){
+        if (zcurr > z0 && polar < M_PI_2)
+          num++;
+        else if (zcurr < z0 && polar > M_PI_2)
+          num++;
+        else if (fabs(zcurr - z0) < 1.e-10 && fabs(polar - M_PI_2) < 1.e-10)
+          num++;
+      }
+      else if (azim > M_PI && ycurr < y0){
+        if (zcurr > z0 && polar < M_PI_2)
+          num++;
+        else if (zcurr < z0 && polar > M_PI_2)
+          num++;
+        else if (fabs(zcurr - z0) < 1.e-10 && fabs(polar - M_PI_2) < 1.e-10)
+          num++;        
+      }
     }
 
     /* There are two intersections */
     else {
       xcurr = x0;
       ycurr = (-b + sqrt(discr)) / (2 * a);
-      points[num].setCoords(xcurr, ycurr);
-      if (angle < M_PI && ycurr > y0)
-        num++;
-      else if (angle > M_PI && ycurr < y0)
-        num++;
+      zcurr = z0 + sqrt(pow(ycurr - y0, 2.0) + pow(xcurr - x0, 2.0))
+        * tan(M_PI_2 - polar);
+      points[num].setCoords(xcurr, ycurr, zcurr);
+      if (azim < M_PI && ycurr > y0){
```

Comment:
Spacing. I'll stop with these comments - just search for ``){`` and replace with ``) {``